timesketch/lib/analyzers/contrib/misp_analyzer.py

```diff
@@ -0,0 +1,151 @@
+        self.misp_api_key = kwargs.get("misp_api_key")
+        self.total_event_counter = 0
+        self.request_set = set()
+
+    @staticmethod
+    def get_kwargs():
+        """Get kwargs for the analyzer.
+
+        Returns:
+            Info to connect to MISP.
+        """
+
+        misp_url = current_app.config.get("MISP_URL")
+        misp_api_key = current_app.config.get("MISP_API_KEY")
+
+        if not misp_api_key or not misp_url:
+            logger.error("MISP conf not found")
+            return []
+
+        matcher_kwargs = [{"misp_url": misp_url, "misp_api_key": misp_api_key}]
+        return matcher_kwargs
+
+    def get_misp_attributes(self, event, attr):
+        """Search event on MISP.
+
+        Returns:
+            List of matching MISP attibutes.
+        """
+        results = requests.post(
+            f"{self.misp_url}/attributes/restSearch/",
+            data={"returnFormat": "json", "value": event, "type": attr},
+            headers={"Authorization": self.misp_api_key},
+            verify=False,
+        )
+
+        if results.status_code != 200:
+            logger.error("Error with MISP url.")
+            return []
+        if "name" in results.json():
+            if "Authentication failed." in results.json()["name"]:
+                logger.error("Bad API key. Please change it.")
+                return []
+        if not results.json()["response"]["Attribute"]:
+            return []
+
+        self.total_event_counter += 1
+
+        return results.json()["response"]["Attribute"]
+
+    def mark_event(self, event, result, attr):
+        """Anotate an event with data from MISP result.
+
+        Add a comment to the event.
+        """
+
+        msg = "Event that match files:   "
+        for misp_attr in result:
+            info = misp_attr["Event"]["info"]
+            id_event = misp_attr["Event"]["id"]
+            msg += f'"Event info": "{info}"'
+            msg += f"- \"url\": {self.misp_url + '/events/view/' + id_event} || "
+
+        event.add_comment(msg)
+        event.add_tags([f"MISP-{attr}"])
+        event.commit()
+
+    def query_misp(self, query, attr, timesketch_attr):
+        """Get event from timesketch, request MISP and mark event."""
+        events = self.event_stream(query_string=query, return_fields=[timesketch_attr])
+        create_a_view = False
+        for event in events:
+            loc = event.source.get(timesketch_attr)
+            if attr == "filename":
+                loc = ntpath.basename(loc)
+                if not loc:
+                    _, loc = ntpath.split(event.source.get(timesketch_attr))
+
+            if not loc in self.request_set:
+                result = self.get_misp_attributes(loc, attr)
+                if result:
+                    create_a_view = True
+                    self.mark_event(event, result, attr)
+                self.request_set.add(loc)
+            else:
+                self.mark_event(event, result, attr)
```

Comment:
Same note as for hashlookup - won't this tag all events the second time they show up, regardless of whether there are results or not? Also, here `result` is undefined.

Reply:
Fixed in 18325158858d824839fb5825bcdc6e1bd6c14552.